src/render.rs

```diff
@@ -345,6 +364,81 @@ impl<'a> Renderer<'a> {
         Ok(output.trim_right().to_string())
     }
 
+    fn render_macro(&mut self, call_node: Node) -> TeraResult<String> {
+        if let MacroCall {namespace, name: macro_name, params: call_params} = call_node {
+            // We need to find the active namespace in Tera if `self` is used
+            // Since each macro (other than the `self` ones) pushes its own namespace
+            // to the stack when being rendered, we can just lookup the last namespace that was pushed
+            // to find out the active one
+            let active_namespace = match namespace.as_ref() {
+                "self" => {
+                    // TODO: handle error if we don't have a namespace
+                    // This can only happen when calling {{ self:: }} outside of a macro afaik
+                    self.macro_namespaces
+                        .last()
+                        .expect("Open an issue with a template sample please (mention `self namespace macro`)!")
+                        .to_string()
+                },
+                _ => {
+                    // TODO: String doesn't have Copy trait, can we avoid double cloning?
+                    self.macro_namespaces.push(namespace.clone());
+                    namespace.clone()
+                }
+            };
+
+            // We get our macro definition using the namespace name we just got
+            let macro_definition = self.macros
+                .get(&active_namespace)
+                .and_then(|m| m.get(&macro_name).cloned());
+
+            if let Some(Macro {body, params, ..}) = macro_definition {
+                // fail fast if the number of args don't match
+                if params.len() != call_params.len() {
+                    let params_seen = call_params.keys().cloned().collect::<Vec<String>>();
+                    return Err(MacroCallWrongArgs(macro_name, params, params_seen));
+                }
+
+                // We need to make a new context for the macro from the arguments given
+                // Return an error if we get some unknown params
+                let mut context = HashMap::new();
+                for (param_name, exp) in call_params.clone() {
```

Comment:
Ah, that's a bummer. All of the `eval` expression should allow references as well. Like taking in a T: Borrow<parser::Node> instead of just a parser::Node. But, meh.